ddtrace/tracer/option.go

```diff
@@ -135,6 +135,75 @@ func defaults(c *config) {
 	}
 }
 
+// fallbackToTags checks if env, service, and version are set. If not,
+// it tries to load their values from globalTags.
+func (c *config) fallbackToTags() {
+	if c.env == "" {
+		if v, ok := c.globalTags["env"]; ok {
+			if e, ok := v.(string); ok {
+				c.env = e
+			}
+		}
+	}
+	if c.serviceName == "" {
+		if v, ok := c.globalTags["service"]; ok {
+			if s, ok := v.(string); ok {
+				c.serviceName = s
+				globalconfig.SetServiceName(s)
+			}
+		}
+	}
+	if c.version == "" {
+		if v, ok := c.globalTags["version"]; ok {
+			if ver, ok := v.(string); ok {
+				c.version = ver
+			}
+		}
+	}
+}
+
+// defaults ensures that certain fields in the tracer are set to their default
+// values, if a value has not been configured.
+func (c *config) defaults() {
+	if c.serviceName == "" {
+		c.serviceName = filepath.Base(os.Args[0])
+	}
+	if c.transport == nil {
+		c.transport = newTransport(c.agentAddr, c.httpClient)
+	}
+	if c.propagator == nil {
+		c.propagator = NewPropagator(nil)
+	}
+	if c.logger != nil {
+		log.UseLogger(c.logger)
+	}
+	if c.debug {
+		log.SetLevel(log.LevelDebug)
+	}
+	if c.statsd == nil {
+		client, err := statsd.New(c.dogstatsdAddr, statsd.WithMaxMessagesPerPayload(40), statsd.WithTags(statsTags(c)))
+		if err != nil {
+			log.Warn("Runtime and health metrics disabled: %v", err)
+			c.statsd = &statsd.NoOpClient{}
+		} else {
+			c.statsd = client
+		}
+	}
+}
+
+func newConfig(opts ...StartOption) *config {
+	c := new(config)
+	c.sampler = NewAllSampler()
+	c.agentAddr = defaultAddress
+	c.loadEnv()
+	for _, fn := range opts {
+		fn(c)
+	}
+	c.fallbackToTags()
+	c.defaults()
```

Comment:
This does not look right. Defaults are values we start with, not to end with.

Reply:
In this case, we apply default values at the end, if certain fields are unset. That's how the code behaved before, but I can try moving the defaults to the beginning.